core.d.ts

```diff
@@ -0,0 +1,1237 @@
+
+  /**
+   * Check if the request is fresh, aka
+   * Last-Modified and/or the ETag
+   * still match.
+   */
+  readonly fresh: boolean;
+
+  /**
+   * Check if the request is stale, aka
+   * "Last-Modified" and / or the "ETag" for the
+   * resource has changed.
+   */
+  readonly stale: boolean;
+
+  /**
+   * Check if the request was an _XMLHttpRequest_.
+   */
+  readonly xhr: boolean;
+
+  // body: { username: string; password: string; remember: boolean; title: string; };
+  body: ReqBody;
+
+  // cookies: { string; remember: boolean; };
+  cookies: any;
+
+  method: string;
+
+  params: P;
+
+  query: ReqQuery;
+
+  route: any;
+
+  signedCookies: any;
+
+  originalUrl: string;
+
+  url: string;
+
+  baseUrl: string;
+
+  app: Application;
+
+  /**
+   * After middleware.init executed, Request will contain res and next properties
+   * See: express/lib/middleware/init.js
+   */
+  res?: Response<ResBody, LocalsObj> | undefined;
+  next?: NextFunction | undefined;
+}
+
+export interface MediaType {
+  value: string;
+  quality: number;
+  type: string;
+  subtype: string;
+}
+
+export type Send<ResBody = any, T = Response<ResBody>> = (body?: ResBody) => T;
+
+export interface SendFileOptions extends SendOptions {
+  /** Object containing HTTP headers to serve with the file. */
+  headers?: Record<string, unknown>;
+}
+
+export interface DownloadOptions extends SendOptions {
+  /** Object containing HTTP headers to serve with the file. The header `Content-Disposition` will be overridden by the filename argument. */
+  headers?: Record<string, unknown>;
+}
+
+export interface Response<
+  ResBody = any,
+  LocalsObj extends Record<string, any> = Record<string, any>,
+  StatusCode extends number = number,
+> extends http.ServerResponse, Express.Response {
+  /**
+   * Set status `code`.
+   */
+  status(code: StatusCode): this;
+
+  /**
+   * Set the response HTTP status code to `statusCode` and send its string representation as the response body.
+   * @link http://expressjs.com/4x/api.html#res.sendStatus
+   *
+   * Examples:
+   *
+   *    res.sendStatus(200); // equivalent to res.status(200).send('OK')
+   *    res.sendStatus(403); // equivalent to res.status(403).send('Forbidden')
+   *    res.sendStatus(404); // equivalent to res.status(404).send('Not Found')
+   *    res.sendStatus(500); // equivalent to res.status(500).send('Internal Server Error')
+   */
+  sendStatus(code: StatusCode): this;
+
+  /**
+   * Set Link header field with the given `links`.
+   *
+   * Examples:
+   *
+   *    res.links({
+   *      next: 'http://api.example.com/users?page=2',
+   *      last: 'http://api.example.com/users?page=5'
+   *    });
+   */
+  links(links: any): this;
+
+  /**
+   * Send a response.
+   *
+   * Examples:
+   *
+   *     res.send(new Buffer('wahoo'));
+   *     res.send({ some: 'json' });
+   *     res.send('<p>some html</p>');
+   *     res.status(404).send('Sorry, cant find that');
+   */
+  send: Send<ResBody, this>;
```

Comment:
Re: https://github.com/expressjs/express/issues/6385

What are folk's thoughts on taking this PR as an opportunity to slightly tighten the types around function input like this? In the actual code, a fixed set of "types" are documented but is not reflected in the TS types.

e.g.

https://github.com/expressjs/express/blob/0bb00e19068e9d95afcebb1821ab1eec5c1430aa/lib/response.js#L110-L123

Perhaps such changes should land in a follow-on PR but it seemed worth proposing in this context.